File: collect_de_fond_SpringBoot/collectdefond/src/main/java/com/example/collectdefond/Controllers/DonationController.java
package com.example.collectdefond.Controllers;

import com.example.collectdefond.Models.Campaign;
import com.example.collectdefond.Models.Donation;
import com.example.collectdefond.Models.Donation.PaymentMethod;
import com.example.collectdefond.ServiceImpl.DonationServImp;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.HashMap;

@RestController
@RequestMapping("/api/donations")
@CrossOrigin(origins = "*")
public class DonationController {

    @Autowired
    private DonationServImp donationServ;

    @GetMapping("/total")
    public ResponseEntity<Map<String, Object>> getTotalDonations() {
        try {
            long totalDonations = donationServ.getTotalDonations();
            Map<String, Object> response = new HashMap<>();
            response.put("total", totalDonations);
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            Map<String, Object> errorResponse = new HashMap<>();
            errorResponse.put("error", "Erreur lors de la récupération du total des donations");
            errorResponse.put("message", e.getMessage());
            return new ResponseEntity<>(errorResponse, HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

    @PostMapping("/donate")
    public ResponseEntity<?> donate(@RequestBody Map<String, Object> payload) {
        try {
            // Validation des champs requis
            if (!payload.containsKey("donorName") || !payload.containsKey("donorEmail") ||
                    !payload.containsKey("amount") || !payload.containsKey("paymentMethod") ||
                    !payload.containsKey("campaign_id")) {
                return new ResponseEntity<>("Tous les champs requis doivent être remplis",
                        HttpStatus.BAD_REQUEST);
            }

            Donation donation = new Donation();

            // Information du donateur
            donation.setDonorName((String) payload.get("donorName"));
            donation.setDonorEmail((String) payload.get("donorEmail"));

            // Montant
            try {
                double amount = Double.parseDouble(payload.get("amount").toString());
                if (amount <= 0) {
                    return new ResponseEntity<>("Le montant doit être supérieur à 0",
                            HttpStatus.BAD_REQUEST);
                }
                donation.setAmount(amount);
            } catch (NumberFormatException e) {
                return new ResponseEntity<>("Montant invalide", HttpStatus.BAD_REQUEST);
            }
            try {
                PaymentMethod paymentMethod = PaymentMethod.valueOf(
                        ((String) payload.get("paymentMethod")).toUpperCase());
                donation.setPaymentMethod(paymentMethod);
                if (payload.containsKey("cardNumber")) {
                    donation.setCardNumber((String) payload.get("cardNumber"));
                }
            } catch (IllegalArgumentException e) {
                return new ResponseEntity<>("Méthode de paiement invalide",
                        HttpStatus.BAD_REQUEST);
            }

            // Campagne
            try {
                Campaign campaign = new Campaign();
                campaign.setId(Long.valueOf(payload.get("campaign_id").toString()));
                donation.setCampaign(campaign);
            } catch (NumberFormatException e) {
                return new ResponseEntity<>("ID de campagne invalide",
                        HttpStatus.BAD_REQUEST);
            }
            donation.setStatus(Donation.Status.valueOf("PENDING"));
            Donation savedDonation = donationServ.donate(donation);
            return new ResponseEntity<>(savedDonation, HttpStatus.CREATED);

        } catch (Exception e) {
            Map<String, String> errorResponse = new HashMap<>();
            errorResponse.put("error", "Erreur lors de la création de la donation");
            errorResponse.put("message", e.getMessage());
            return new ResponseEntity<>(errorResponse, HttpStatus.BAD_REQUEST);
        }
    }

    @GetMapping("/by-email")
    public ResponseEntity<?> getDonationsByEmail(@RequestParam String email) {
        try {
            if (email == null || email.trim().isEmpty()) {
                return new ResponseEntity<>("L'email ne peut pas être vide",
                        HttpStatus.BAD_REQUEST);
            }

            List<Donation> donations = donationServ.getDonationsByEmail(email);

            if (donations.isEmpty()) {
                Map<String, String> response = new HashMap<>();
                response.put("message", "Aucune donation trouvée pour cet email");
                return new ResponseEntity<>(response, HttpStatus.OK);
            }

            return ResponseEntity.ok(donations);
        } catch (Exception e) {
            Map<String, String> errorResponse = new HashMap<>();
            errorResponse.put("error", "Erreur lors de la récupération des donations");
            errorResponse.put("message", e.getMessage());
            return new ResponseEntity<>(errorResponse, HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }
}

File: Back-end/Controller/donationController.js
const Donation = require('../Models/Donation');
const Campaign = require('../Models/Campaign');



exports.getTotalDonationsCount = async (req, res) => {
  try {
    const totalCount = await Donation.countDocuments();
    res.status(200).json({
      totalDonationsCount: totalCount,
    });
  } catch (error) {
    console.error('Erreur lors de la récupération du nombre total de donations :', error);
    res.status(500).json({
      message: 'Erreur lors de la récupération du nombre total de donations.',
      error: error.message,
    });
  }
};



exports.donate = async (req, res) => {
  try {
    const { donorName, donorEmail, amount, campaign, paymentMethod, cardNumber } = req.body;
    if (!donorName || !donorEmail || !amount || !campaign || !paymentMethod) {
      return res.status(400).json({ message: 'Tous les champs obligatoires doivent être remplis.' });
    }
    const newDonation = new Donation({
      donorName,
      donorEmail,
      amount,
      campaign,
      paymentMethod,
      cardNumber, 
    });
    await newDonation.save();
    const campaignToUpdate = await Campaign.findById(campaign);
    if (!campaignToUpdate) {
      return res.status(404).json({ message: 'Campagne non trouvée' });
    }

    // Incrémenter le montant actuel de la campagne
    campaignToUpdate.currentAmount += amount;
    await campaignToUpdate.save();

    // Vérifier si le montant actuel atteint ou dépasse le montant cible
    if (campaignToUpdate.currentAmount >= campaignToUpdate.targetAmount) {
      await Campaign.findByIdAndDelete(campaign);

      return res.status(200).json({
        message: 'Merci pour votre donation ! La campagne a atteint son objectif et a été supprimée.',
        donation: newDonation,
      });
    }

    // Retourner une réponse avec la donation et la campagne mise à jour
    res.status(201).json({
      message: 'Merci pour votre donation !',
      donation: newDonation,
      campaign: campaignToUpdate,
    });

  } catch (error) {
    console.error('Erreur lors de l\'enregistrement de la donation :', error);
    res.status(500).json({
      message: 'Erreur lors de l\'enregistrement de la donation.',
      error: error.message,
    });
  }
};

exports.getDonations = async (req, res) => {
  const { email } = req.params;

  // Validation de l'email
  if (!email || !/\S+@\S+\.\S+/.test(email)) {
    return res.status(400).json({ message: 'Email invalide' });
  }

  try {
    const donations = await Donation.find({ donorEmail: email });

    if (donations.length === 0) {
      return res.status(404).json({ message: 'Aucune donation trouvée pour cet email.' });
    }

    res.status(200).json(donations);
  } catch (error) {
    console.error('Erreur lors de la récupération des donations :', error);
    res.status(500).json({ message: 'Erreur lors de la récupération des donations.', error });
  }
};

